src/HooglePlus/FilterTest.hs

```diff
@@ -168,4 +168,35 @@ handleAnyException = (`catch` (return . handler . show :: SomeException -> IO (M
   where
     handler ex  | "timeout" `isInfixOf` ex = Nothing
                 | otherwise = (Just . Left. UnknownError) ex
-fmtFunction_ = printf "((%s) :: %s) %s" :: String -> String -> String -> String
\ No newline at end of file
+fmtFunction_ = printf "((%s) :: %s) %s" :: String -> String -> String -> String
+
+checkDuplicates :: SampleResult -> Environment -> RType -> UProgram -> IO (Bool, SampleResult)
+checkDuplicates result env goalType prog =
+  checkDuplicates' result modules sigStr body
+  where
+    (modules, sigStr, body, _) = extractSolution env goalType prog
+
+
+checkDuplicates' :: SampleResult -> [String] -> String -> String -> IO (Bool, SampleResult)
+checkDuplicates' result modules sigStr body =
+  case result of
+    EmptySample ->
+      ((`Sample` []) <$> generateInputs modules funcSig) >>= (\r -> checkDuplicates' r modules sigStr body)
+    Sample inputs results -> do
+      evals <- evalResults inputs funcSig body
+      let isDuplicated = evals `elem` results
+      let results' = if isDuplicated then results else evals:results
+
+      let retVal = (not isDuplicated, Sample inputs results')
+      print retVal
+      return retVal
+
+  where
+    funcSig = (instantiateSignature . parseTypeString) sigStr
+
+    generateInputs :: [String] -> FunctionSignature -> IO [String]
+    generateInputs modules funcSig = replicateM defaultNumChecks (generateTestInput modules funcSig)
+    evalResults inputs funcSig body = mapM f inputs
+
+    f arg =
```

Comment:
`f` gives no indication of what this function does, please give it a more descriptive name.